### ISSUE TITLE
Failed to launch wda because of "Couldn't start port forwarding on port 8100. Please provide a different port using 'wdaLocalPort' capability" without delete session

### ISSUE THREAD

**Issue by KazuCocoa** [appium/appium#13127]
This issue happened the current master.
1.14.2 worked.

## The problem

Appium fails to create a new session with `useNewWDA` after the first launching on a real device.

I mean:

1. send a create session with `useNewWDA: true`
    - ok
2. send a create session with `useNewWDA: true` 
    - without `deleteSession` before this
   - failed because of remaining port 8100 which was used by _1_
   - I guessed iproxy class failed to release 8100 port in `quite()`
      - https://github.com/appium/appium-xcuitest-driver/blob/master/lib/wda/iproxy.js#L42-L48

```
[debug] [BaseDriver] Event 'wdaStartFailed' logged at 1566837388273 (01:36:28 GMT+0900 (Japan Standard Time))
[debug] [XCUITest] Unable to launch WebDriverAgent because of xcodebuild failure: Couldn't start port forwarding on port 8100. Please provide a different port using 'wdaLocalPort' capability. Make sure you follow the tutorial at https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md. Try to remove the WebDriverAgentRunner application from the device if it is installed and reboot the device.
[XCUITest] Quitting and uninstalling WebDriverAgent
[XCUITest] Shutting down sub-processes
[debug] [XCUITest] Removing WDA application from device
[XCUITest] Error: Unable to launch WebDriverAgent because of xcodebuild failure: Couldn't start port forwarding on port 8100. Please provide a different port using 'wdaLocalPort' capability. Make sure you follow the tutorial at https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md. Try to remove the WebDriverAgentRunner application from the device if it is installed and reboot the device.
[XCUITest]     at quitAndUninstall (/Users/kazu/GitHub/appium-xcuitest-driver/lib/driver.js:505:15)
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[debug] [BaseDriver] Event 'newSessionStarted' logged at 1566837388558 (01:36:28 GMT+0900 (Japan Standard Time))
[debug] [W3C] Encountered internal error running command: Error: Unable to launch WebDriverAgent because of xcodebuild failure: Couldn't start port forwarding on port 8100. Please provide a different port using 'wdaLocalPort' capability. Make sure you follow the tutorial at https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md. Try to remove the WebDriverAgentRunner application from the device if it is installed and reboot the device.
[debug] [W3C]     at quitAndUninstall (/Users/kazu/GitHub/appium-xcuitest-driver/lib/driver.js:505:15)
[HTTP] <-- POST /wd/hub/session 500 1716 ms - 1293
```

Another error in step2
```
[debug] [XCUITest] Killing running processes 'xcodebuild.*242bfcc998fd156df0bd0ad1dde8ab8e0a032114' for the device 242bfcc998fd156df0bd0ad1dde8ab8e0a032114...
[XCUITest] xcodebuild exited with code '75' and signal 'null'
[debug] [XCUITest] 'pgrep -ifn xcodebuild.*242bfcc998fd156df0bd0ad1dde8ab8e0a032114' didn't detect any matching processes. Return code: 1
uncaughtException: listen EADDRINUSE: address already in use :::8100
Error: listen EADDRINUSE: address already in use :::8100
    at Server.setupListenHandle [as _listen2] (net.js:1279:14)
    at listenInCluster (net.js:1327:12)
    at Server.listen (net.js:1414:7)
    at iProxy.listen [as start] (/Users/kazu/GitHub/appium/node_modules/appium-xcuitest-driver/lib/wda/iproxy.js:33:23)
    at WebDriverAgent.start [as createIProxy] (/Users/kazu/GitHub/appium/node_modules/appium-xcuitest-driver/lib/wda/webdriveragent.js:249:14)
    at WebDriverAgent.createIProxy [as launch] (/Users/kazu/GitHub/appium/node_modules/appium-xcuitest-driver/lib/wda/webdriveragent.js:228:28)
[debug] [XCUITest] Successfully updated '/Users/kazu/GitHub/appium/node_modules/appium-webdriveragent/WebDriverAgent.xcodeproj/project.pbxproj' with bundle id 'com.trident.WebDriverAgent'
[debug] [XCUITest] Generating xcode config file for orgId '9U4PT397Q4' and signingId 'iPhone Developer'
[debug] [XCUITest] Writing xcode config file to /var/folders/34/2222sh8n27d6rcp7jqlkw8km0000gn/T/2019727-20152-xx57uh.t0rw/appium-temp.xcconfig
```

## Environment

* Appium version (or git revision) that exhibits the issue: master (xcuitest-driver was 7f3e52f03b9726eba8a6173de09db5644a3998f9)
* Last Appium version that did not exhibit the issue (if applicable): 1.14.2
* Desktop OS/version used to run Appium: macOS 10.14.6
* Node.js version (unless using Appium.app|exe): 10.16.0
* Npm or Yarn package manager: npm
* Mobile platform/version under test:
* Real device or emulator/simulator: real device
* Appium CLI or Appium.app|exe:

## Details

https://github.com/appium/appium-xcuitest-driver/blob/master/lib/wda/iproxy.js#L42-L48 has regression?
(I haven't dug much yet)

**Comment by KazuCocoa** [comment]
@umutuzgur
Do you have any idea good about this?

**Comment by mykola-mokhnach** [comment]
Seems like duplicate of https://github.com/appium/appium/issues/13117

**Comment by KazuCocoa** [comment]
mmm.
https://gist.github.com/KazuCocoa/1aaa162cafdd26c0738f4be94a907c55 is the gist in step2.

I inserted some logs which have `=======` in iproxy.js tho.

Someone did not release the port, so next running iproxy failed to establish the connection, it seemed. After killing appium process and re-launching it, it worked once.
`destroy` probably worked since 8100 looked listening as below.

```
kazu$ lsof -i :8100
COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    27795 kazu   25u  IPv6 0x15860d395258063f      0t0  TCP *:xprint-server (LISTEN)
node    27795 kazu   31u  IPv4 0x15860d39492ab3ff      0t0  TCP localhost:61037->localhost:xprint-server (ESTABLISHED)
node    27795 kazu   32u  IPv6 0x15860d39590b6d3f      0t0  TCP localhost:xprint-server->localhost:61037 (ESTABLISHED)
kazu$ lsof -i :8100 # after this.serverSocket.close()
COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    27795 kazu   25u  IPv6 0x15860d395258063f      0t0  TCP *:xprint-server (LISTEN)
kazu$ lsof -i :8100
```

**Comment by KazuCocoa** [comment]
I've added logs in another point as below.

```
    const status = new B((resolve, reject) => {
      this.serverSocket.once('listening', resolve);
      this.serverSocket.once('error', reject);
    });
    this.serverSocket.listen(this.localport);

    log.warn(`=========before status`);
    await status;
    log.warn(`=========after status`);
  }
```

https://github.com/appium/appium-xcuitest-driver/blob/master/lib/wda/iproxy.js#L34-L39

The below log is it. Hm.. the the 2nd step failed because of failing `await status;`...
```
[XCUITest] xcodebuild exited with code '75' and signal 'null'
[debug] [XCUITest] 'pgrep -ifn xcodebuild.*242bfcc998fd156df0bd0ad1dde8ab8e0a032114' didn't detect any matching processes. Return code: 1
[XCUITest] =========before iproxy.start()
[iProxy] =========before status
[XCUITest] Launching WebDriverAgent on the device
[XCUITest] Skipped WDA dependencies resolution according to the provided capabilities
[debug] [XCUITest] Killing running processes 'xcodebuild.*242bfcc998fd156df0bd0ad1dde8ab8e0a032114' for the device 242bfcc998fd156df0bd0ad1dde8ab8e0a032114...
[debug] [XCUITest] 'pgrep -ifn xcodebuild.*242bfcc998fd156df0bd0ad1dde8ab8e0a032114' didn't detect any matching processes. Return code: 1
[XCUITest] =========before iproxy.start()
[iProxy] =========before status
[debug] [BaseDriver] Event 'wdaStartFailed' logged at 1566869664649 (10:34:24 GMT+0900 (Japan Standard Time))
[debug] [XCUITest] Unable to launch WebDriverAgent because of xcodebuild failure: Couldn't start port forwarding on port 8100. Please provide a different port using 'wdaLocalPort' capability. Make sure you follow the tutorial at https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md. Try to remove the WebDriverAgentRunner application from the device if it is installed and reboot the device.
```

**Comment by umutuzgur** [comment]
@KazuCocoa It is happening because of the cached xcodebuild processes. In the past, we killed any process that is associated with the udid of the device but now that we use native nodejs code instead of the iproxy process, we don't have any iproxy process running. We need to kill the iproxy object associated with the xcodebuild to free that port

**Comment by KazuCocoa** [comment]
ah... i see.

Once I call `deleteSession` before step2, it worked well.
With `--session-override`, it did not happen when I called createSession twice without deleteSession between them.

We can guide users to make sure they call `quit` client if someone sayied this issue,

**Comment by umutuzgur** [comment]
@KazuCocoa I will work on a hacky solution to fix it. However, the webdriveragent.js class needs a proper refactor

**Comment by umutuzgur** [comment]
This will fix the issue https://github.com/appium/appium-xcuitest-driver/pull/1053

**Comment by umutuzgur** [comment]
Published the fix for it. This issue should be resolved

**Comment by KazuCocoa** [comment]
This error can happen after the change, unfortunately.
But it improved another waiting behaviour, I think.

**Comment by umutuzgur** [comment]
@KazuCocoa We should just do random port allocation. That would fix the wdaLocalPort issue

**Comment by KazuCocoa** [comment]
Yes, if users do not specify `wdaLocalPort`

**Comment by cosminstirbu** [comment]
I've just installed appium@beta and I'm seeing the same issue.

Has the fix been published to the latest beta?

**Comment by KazuCocoa** [comment]
Current beta has appium/appium-xcuitest-driver#1053 , but no random port selection such as the above comment.
How did you reproduce your issue? The full appium log is helpful.

My original issue, https://github.com/appium/appium/issues/13127#issue-485338713 , can avoid this error with calling `driver.quit` (a.k.a delete session) expectedly before starting a new session (without `--session-override` server option)

**Comment by umutuzgur** [comment]
There is not much we can do without the random port allocation or `wdaLocalPort` should be passed as capability. The main problem we have is that we always try to allocate 8100 unless `wdaLocalPort` is provided. Of course, 2 sessions can't allocate the same port and it doesn't make sense that a second session kills the processes of the first one. @KazuCocoa the fix for random port allocation will me difficult because our WDA process caching mechanism is pretty messed up at the moment. We need to refactor that one first

**Comment by KazuCocoa** [comment]
Yeah, right. > the fix for...

**Comment by cosminstirbu** [comment]
So I've had a closer look at this because I was a bit confused since I didn't understand the exact reproduction steps.

During test development we run our tests on a real device by connecting to an instance of Appium server that was manually started , and while we do call reset on the driver (which deletes the previous session) between each test, at then end of our test execution, we don't delete the last session, so I assume WDA keeps the session running.

So next time we run our tests, the previous session is still alive on WDA, so that's when we get this response.

I assume a fix for this would be to kill the session when our all of our tests end.

**Comment by KazuCocoa** [comment]
> we don't delete the last session, so I assume WDA keeps the session running.

Yes, but it is xcodebuild process to run WDA than WDA itself.
xcodebuild probably listens to the port, so `Couldn't start port forwarding on port 8100` happens.

> I assume a fix for this would be to kill the session when our all of our tests end.

Yes. It is a solid way to avoid this issue.

----

Or you can communicate with WDA via WiFi network using `webDriverAgentUrl` or `wdaBaseUrl` and `wdaLocalPort`.

e.g.
`wdaBaseUrl: 'http://192.168.0.1'` and `wdaLocalPort: '8100'` => xcodebuild communicate with the device via `8100` port forwarding. Appium communicate with WDA running on the device via `http://192.168.0.1:8100`

**Comment by cosminstirbu** [comment]
I want to avoid hardcoding any IP address as our device is used for manual testing during the day and we might switch it to several different networks.

With the current beta, if we would pass a random open port to `wdaLocalPort` each time we create a new session would it be enough to avoid this issue? Or we first need https://github.com/appium/appium-xcuitest-driver/pull/1056 to be merged? Or even after that, passing a random open port to `wdaLocalPort` wouldn't be enough?

The issue is that we have an on premise CI agent, and for some reason even after we have introduced the deletion of the session after the last test, the issue still reproduces if we kick off a new job (even though there is no `xcodebuild`, `node` instances after the job ends).

I admit I don't fully understand the architecture, so I'm not sure if the port error means that the issue is with the machine where Appium server is running or with the device.

It seems that you're suggesting that `xcodebuild` might keep the port busy, but I don't have `xcodebuild` nor `node` showing up in Activity Monitor (My Processes) so I'm not sure what the issue is.

If I delete WDA from the device, then the tests run fine, until they finish running and I trigger them once again, without deleting the WDA.

**Comment by cosminstirbu** [comment]
Actually I take it back, even after I have deleted WDA from the device, the issue still reproduced.

**Comment by cosminstirbu** [comment]
For now I'll rollback to 1.14.0 as I've actually tried the beta version due to this issue https://github.com/appium/appium/issues/13134

For some reason 1.14.0 doesn't have the issue linked above. I'll keep monitoring this issue in case there are any new changes in the beta track that I can test out.

**Comment by KazuCocoa** [comment]
> if we would pass a random open port to wdaLocalPort each time we create a new session would it be enough to avoid this issue?

It is enough, I think.
Using the same port is this issue. Listened port by xcodebuild will expire in a couple of min after releasing the port by Appium (node). Thus, pick a random port every time is pretty enough.

**Comment by KazuCocoa** [comment]
appium/appium-xcuitest-driver#1053 is available as `appium@beta`

**Comment by cosminstirbu** [comment]
Installed the latest beta and switched to passing a random available port to wdaLocalPort and everything works as expected.

Also the reset call doesn't time out either.

Thanks.

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Can we close this issue after your recent fix was merged?

**Comment by KazuCocoa** [comment]
Ha, yes, I think for now.
My PR was one way to avoid this issue.


Summary:

1. Call `delete session` before running a new session so that Appium can finish all deleteSession process properly
2. Set random available `wdaLocalPort` by the user side
3. Use `webDriverAgentUrl` to handle WDA by user side (Appium skip run xcodebuild against WDA)
4. Set `wdaBaseUrl` and `wdaLocalPort` so that Appium can communicate with WDA via local network (no via forward proxy)

1 and 2 are easy way.

**Comment by raysurc** [comment]
> Installed the latest beta and switched to passing a random available port to wdaLocalPort and everything works as expected.
> 
> Also the reset call doesn't time out either.
> 
> Thanks.

How do you pass a random available port to wdalocalPort?

**Comment by cosminstirbu** [comment]
With the .NET client you can use:

`        
private static int GetAvailablePort()
        {
            var listener = new TcpListener(IPAddress.Loopback, 0);

            listener.Start();
            int port = ((IPEndPoint)listener.LocalEndpoint).Port;
            listener.Stop();

            return port;
        }
`

I'm sure something similar can be done for the other clients as well (Java, etc)
